test/hotspot/gtest/nmt/test_nmt_treap.cpp

```diff
@@ -145,6 +146,49 @@ class NMTTreapTest : public testing::Test {
     test(5.0f);
     test(0.0f);
   }
+
+  void upsert_performance_test() {
+    auto n_upsert = [&](int n) {
+      TreapCHeap<int, int, Cmp> treap;
+      double st = os::elapsedTime();
+      for (int i = 0; i < n; i++) {
+        int a = (os::random() % n) * 100;
+        treap.upsert(a, 0);
+      }
+      double d = os::elapsedTime() - st;
+      return d;
+    };
+    const int N1 = 1000;
+    const int N2 = 10000;
+
+    double d1 = n_upsert(N1);
+    double d2 = n_upsert(N2);
+    tty->print_cr("d1: %lf, d2: %lf, d2/d1: %lf", d1, d2, d2 / d1);
+    EXPECT_LT((int)(d2 / d1), N2 / N1);
+  }
```

Comment:
1000 and 10_000 elements are both quite small, isn't one measurement prone to being slower than expected due to unforeseen circumstances? For example, if `malloc` performs a heavy allocation (by mapping in a few new  pages) during the 10_000 elements insertion then that may stall enough that the `EXPECT_LT` fails. We are also bound to only performing the test once.

Is your thinking that these perf tests should be kept in, or should they be removed before integrating?

Reply:
Generally, I think it is necessary to have performance tests to verify if any future changes do not degrade the performance.
The approach of stress testing can be agreed between us, but the tests have to exist.
In this approach, the input size is scaled N times and we expect the execution time grows as O(log(N)) which is much less than N.
This test fails and we need to have a justification for it. If approach is invalid or not correct implemented, we fix it. But the expectations remains the same.

Reply:
Why would the execution time grow logarithmically when we do linearly more work? When we run this with `N2` we will perform `10_000 * log(10_000, 2)` units of work, and for `N1` we will perform `1_000 * log(1_000, 2)` units of work, approximately. A closer approximation is `O(log(1)) + O(log(2)) + ... + O(log(n))` for `n = N2` or `n = N1`.

Let's calculate that:

```
>>> import math
>>> def time_it(n):
...     t = 0
...     for i in range(1, n):
...             t  = t + math.log(i)
...     return [t, 3*t] # Approximate best and worst case
... 
>>> time_it(1000)
[5905.220423209189, 17715.661269627566]
>>> time_it(10_000)
[82099.71749644217, 246299.15248932652]
>>> def compare(a, b):
...     ab, aw = a
...     bb, bw = b
...     return [ bb / ab, bw / aw ]
... 
>>> compare(time_it(1000), time_it(10_000))
[13.902904821938064, 13.902904821938066]
>>> # Ouch, that's outside of our linear bound!
```

What I said previously also applies, especially the performance of `malloc` will impact this I suspect.

Reply:
It is considered that `malloc` or other external events are the same for two cases. If we know that there might be some noise for one or another, we should check and disable them. This is the approach I have talked. How can we avoid noise from `malloc` side?

Reply:
When it is said that an algorithm has the log(n) time-complexity, it means that if the input grows n times, the times grows log(n) times. The tree data-structure has log(n) time-complexity. VMATree may have not exactly log(n) response times due to self-balancing costs. But it is still expected to be less than O(n).

Reply:
Hi Afshin, could we take a step back and do some asymptotic time complexity analysis of this problem?

The test is measuring the following code:

```c++
for (int i = 0; i < n; i++) {
  int a = (os::random() % n) * 100;
  treap.upsert(a, 0);
}
```

So this algorithm is the tme complexity which we are trying to understand. First, let's simplify the code slightly:

```c++
auto f = [&](auto n) { int a = (os::random() % n) * 100;  treap.upsert(a, i); };
for (int i = 0; i < n; i++) {
  f(i);
}
```

Clearly, `f(n)` is executed `n` times, agreed? Then the time complexity of the whole program must be `O(n*f(n))`, correct? It's the time complexity of `f(n)` performed `n` times.

Let's replace `f` with something else to see if we can understand the time complexity of the whole code snippet again.

```c++
int arr[n];
auto f = [&](auto n) { arr[n] = 0; };
for (int i = 0; i < n; i++) {
  f(i);
}
```

Now, we both agree that assigning to an array has time complexity `O(1)`, correct? Then, if we fill that in in our expression `O(n * f(n))` we receive `O(n * 1) = O(n)`, correct? In other words, the time complexity of the algorithm the test is measuring is *linear*, and we ought to expect that with an array the time taken for the array should be 10x longer with 10k elements as compared to 1k elements.

OK, now let's *assume* that `f(n)` has time complexity `O(log n)`, then shouldn't the algorithm we're measuring have time complexity `O(n * log n)`, that is actually *slower* than `O(n)`.

In conclusion: if `treap.upsert()` has time complexity `O(log n)` then the whole algorithm should have time complexity `O(n * log n)` and the measurements we're seeing are as expected.

Have I missed anything or made any mistakes? Please let me know.

Reply:
The big O is a _notation_ and is not a math function. So `O(a * b)` is not always same as `O(a) * O(b)`.
Stick to this _definition_: "when an algorithm has time-complexity of O(`g(n)`), it means if the input grows `n` times the time of executing the algorithm grows `g(n)` times." Where `g()` is `log()` in our case.
IOW, the big O answers the following question:
$t_1$ = time for running `f()` for $n_1$ items
$t_2$ = time for running `f()` for $n_2$ items
if we know $\Large{\frac{n_2}{n_1} = n}$ what is expected value of $t_2$?

A detailed description can be found [here](https://en.wikipedia.org/wiki/Big_O_notation).